### PR TITLE
Categorized tenants

### PR DESCRIPTION
**Pull request by hishamco**
Addresses #10402

Screenshot 2021-10-29 182651

- [x] Sorting
- [x] Filtering
- [x] Pagination
- [x] Searching

### CODE REVIEW

**Comment by hishamco** [comment]
Screenshot 2021-10-29 213421

File: src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs
Comment (jtkech) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs:45]: `ShellSettings` holds 2 `ShellConfiguration` config sources

https://github.com/OrchardCMS/OrchardCore/blob/dbd83c6f226c15db5e6ba82b0e2c1609aa693d77/src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs#L20-L24

For both of them, a value can be pre-configured e.g. through the main `appsettings.json`, but when a value is added / updated at runtime, e.g. through a setup or while editing a tenant, a value updated in `_configuration` will be persisted in the tenant specific `appsettings.json` under the tenant folder (can be configured to be in an azure blob or database), a value updated in `_settings` will be persisted in the `tenants.json` (shared by all tenants) directly under the `App_Data` folder (can be configured to be elsewhere, in an azure blob or database).

The `tenants.json` file (related to `_settings`) is intended to hold the minimum set of values (not already pre-configured) required to init a tenant, the other values held by tenant specific sources (related to `_configuration`) being used to lazily build a tenant.

So here for `Category` better to not use a dedicated field using `_settings`, but instead the following indexer that uses `_configuration`, as we already do e.g. for the tenant `Description`.

https://github.com/OrchardCMS/OrchardCore/blob/dbd83c6f226c15db5e6ba82b0e2c1609aa693d77/src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs#L69-L74

So nothing to do here, and in the other place just do what we already did for the tenant `Description`.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs:45]: Thanks for the info @jtkech, I will use indexer instead

**Comment by deanmarcussen** [comment]
I don't like `Not Categorized`. For any existing user it would force them to go through and categorise all their tenants, just to get rid of an odd bit of text.

Categories (if we do it), should be optional.
So just use `String.Empty`, and `None` in the filter list (like we do elsewhere`), and don't show anything if it is `String.Empty`

**Comment by hishamco** [comment]
> And equally don't show the filter if there are no categories.

So, how can we list the tenants that they don't have category?

> And equally don't show the filter if there are no categories.

+1

**Comment by deanmarcussen** [comment]
> So, how can we list the tenants that they don't have category?

you would have an All. It would be the default.

We would never have a filter that forces people to put things into categories.

**Comment by hishamco** [comment]
> you would have an All. It would be the default.

No make sense, coz when you said None before ;) I will update the PR soon ...

**Comment by hishamco** [comment]
Any updates on this?

**Comment by sebastienros** [comment]
Can you update the screenshots if you removed the "not categorized" ?

**Comment by hishamco** [comment]
Tenants categories

**Comment by hishamco** [comment]
Please don't merge after I double check the merge conflict

**Comment by deanmarcussen** [comment]
@hishamco you have missed my previous review.

Where I said that `None` shouldn't be showing.

If people aren't wanting to categorise their tenants (I can't think of any category I'd give mine, they are all of the same type), then having a ranom None tag showing up makes no sense.

i.e. Ok, with categorising, not ok with forcing it on users.

**Comment by hishamco** [comment]
Oops, seems I forgot this one, I will update the PR soon ...

**Comment by hishamco** [comment]
If there's no other comment, then the PR is ready to merge

File: src/OrchardCore.Modules/OrchardCore.Tenants/ViewModels/AdminIndexViewModel.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/ViewModels/AdminIndexViewModel.cs:72]: If we're going to rename this is should be `TenantState`  or `TenantsState` not status.

Otherwise looks good. 
Can you provide a new graphic please.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/ViewModels/AdminIndexViewModel.cs:72]: > Can you provide a new graphic please.

For? I already attached the latest screenshoort [here](https://github.com/OrchardCMS/OrchardCore/pull/10586#issuecomment-991077437)

Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Tenants/ViewModels/AdminIndexViewModel.cs:72]: Because the last screenshot didn't make sense, and I had to comment on it requesting more changes.

So I can see what you have done this time, and make sure it makes sense, please include a new screenshot

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/ViewModels/AdminIndexViewModel.cs]: Sorry, seems I removed the `None` badge as you suggested earlier. I need to update the screenshot

**Comment by hishamco** [comment]
Tenants categories

**Comment by hishamco** [comment]
@deanmarcussen anything left here or can we approve then merge?

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:128]: Since they are inserted (on the edit screen) manually I think you should use ordinal ignore case here.
Otherwise alpha and Alpha are different categories which would be odd

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:185]: This is only selected:true when it’s all or empty isn’t it?

so here and in the above you need to do a comparison against the currently selected value to set the is selected property

**Review by deanmarcussen — COMMENTED** [review]
Two minor issues to fix. Otherwise looks good

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:185]: I think no need to ignore the comparison casing here because the `DefaultCategory` is not editable. Regarding the check you are right need to be changed

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:181]: here is where you need to add the comparison, otherwise when these items are selected they will never be shown as selected.

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:186]: You don't need `DefaultCategory` at all. when it is Empty it is the default category. Just remove it

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:181]: I didn't get it, I'm just checking if the category is `null` or empty, could you please write a snippet of what you mean

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:186]: Right

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:181]:
```csharp
model.Options.TenantsCategories = allSettings
    .GroupBy(t => t["Category"])
    .Where(t => !String.IsNullOrEmpty(t.Key) && !t.Key.Equals(DefaultCategory, StringComparison.OrdinalIgnoreCase))
    .Select(t => new SelectListItem(t.Key, t.Key))
    .ToList();
```

Is it what you mean?

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:186]:
```suggestion
                String.Empty,
```

And remove `DefaultCategory` - it works, and we do this everywhere else.

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:181]:
```
model.Options.TenantsCategories = allSettings
    .GroupBy(t => t["Category"])
    .Where(t => !String.IsNullOrEmpty(t.Key))
    .Select(t => new SelectListItem(t.Key, t.Key, String.Equals(model.Category, t.Key, StringComparison.OrdinalIgnoreCase)))
    .ToList();
```

where model.Category = Selected Category, and you set whether the `SelectListItem` is selected or not

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:186]: Same thing

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:181]: Go it, you want to set th select property, I will update the PR

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:187]:
```suggestion
                selected: String.IsNullOrEmpty(options.Category));
```

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:125]:
```suggestion
            if (!String.IsNullOrWhiteSpace(options.Category))
```

File: src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:29]: Remove

**Review by deanmarcussen — COMMENTED** [review]
You forgot to remove the default category - it's unnecesary

**Comment by hishamco** [comment]
Seems I forgot to push a local commit from awhile ;)

**Review by deanmarcussen — APPROVED** [review]

**Pull request merged**
